### PR TITLE
remove zero Balance or empty NFTGallery

### PR DESCRIPTION
**Pull request by 0xpause**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number:     #20

## What is the new behavior?

- A script function to remove zero balance Token
- A script function to remove empty NFTGallery

### CODE REVIEW

File: integration-tests/account/remove_zero_balance.exp [hunk 1]
```diff
@@ -0,0 +1,9 @@
+processed 3 tasks
+
+task 2 'run'. lines 5-24:
+{
+  "gas_used": 0,
+  "status": {
+    "Discard": 14
```
Comment (jolestar) [inline integration-tests/account/remove_zero_balance.exp:7]: expect status should be `EXECUTED`

File: integration-tests/account/remove_zero_balance.move [hunk 2]
```diff
@@ -0,0 +1,24 @@
+//# init -n dev
+
+//# faucet --addr alice --amount 0
```
Comment (jolestar) [inline integration-tests/account/remove_zero_balance.move:3]: amount 0 will cause no gas fee for executing the transaction, and the transaction will be discard

File: integration-tests/account/remove_zero_balance.move [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        let coin = Token::zero<STC>();
+        Account::deposit_to_self<STC>(&account, coin);
+
+        Account::remove_zero_balance<STC>(&account);
+
+        Account::set_auto_accept_token(&account, false);
+        assert!(!Account::is_accept_token<STC>(addr), 101);
+    }
+}
+// check: EXECUTED
```
Comment (jolestar) [inline integration-tests/account/remove_zero_balance.move:24]: the check statement is no longer needed. We check the transaction status with the exp file

File: integration-tests/nft/test_gallery.exp [hunk 4]
```diff
@@ -95,3 +95,21 @@ task 15 'run'. lines 236-250:
...
+      "MoveAbort": [
+        {
+          "Module": {
+            "address": "0x00000000000000000000000000000001",
+            "name": "Vector"
+          }
+        },
+        3
+      ]
+    }
```
Comment (jolestar) [inline integration-tests/nft/test_gallery.exp:113]: expect status should be Executed, we need to figure out why to abort at Vecotor module.

File: integration-tests/nft/test_gallery.move [hunk 5]
```diff
@@ -247,4 +247,23 @@ fun main(sender: signer) {
...
+    use StarcoinFramework::NFTGallery;
+    use StarcoinFramework::Signer;
+
+    fun main(sender: signer) {
+        let addr: address = Signer::address_of(&sender);
+
+        NFTGallery::accept<AnyNFT, AnyNFTBody>(&sender);
+        assert!(NFTGallery::is_accept<AnyNFT, AnyNFTBody>(addr), 1);
+
+        NFTGallery::remove_empty_gallery<AnyNFT, AnyNFTBody>(&sender);
```
Comment (jolestar) [inline integration-tests/nft/test_gallery.move:264]: Maybe the NFTGallery<AnyNFT, AnyNFTBody> is not empty here, so it aborts at Vector.

File: sources/Account.move [hunk 6]
```diff
@@ -993,6 +993,18 @@ module Account {
...
+    /// Remove zero Balance
+    public fun remove_zero_balance<TokenType: store>(account: &signer) acquires Balance {
+        let addr: address = Signer::address_of(account);
+        let Balance<TokenType> { token } = move_from<Balance<TokenType>>(addr);
+        Token::destroy_zero<TokenType>(token);
+    }
+
+    spec remove_zero_balance {
+        let addr = Signer::address_of(account);
+        aborts_if !exists<Balance<CoinType>>(addr);
```
Comment (whst) [inline sources/Account.move:1006]: An additional postcondition can be added:
`ensures !exists<Balance<CoinType>>(addr);`

File: integration-tests/account/remove_zero_balance.move (hunk 3 above)
Comment (0xpause) [reply integration-tests/account/remove_zero_balance.move:24]: I am not sure what you mean. You mean the `assert` statement is no longer needed?

Comment (jolestar) [reply integration-tests/account/remove_zero_balance.move]: "// check: EXECUTED"  is no longer needed.

**Review by jolestar — APPROVED** [review]
LGTM

**Comment by jolestar** [comment]
Please run ./script/build.sh and commit the document's change.

**Comment by jolestar** [comment]
@0xpause I have rebased the branch, you need to force update the branch.

File: integration-tests/account/remove_zero_balance.move (hunk 3 above)
Comment (jolestar) [reply integration-tests/account/remove_zero_balance.move]: Maybe keeping the "// check: EXECUTED" statement can help read the code. It is ok.

**Comment by 0xpause** [comment]
> Please run ./script/build.sh and commit the document's change.

Done.

**Pull request merged**